In [1]:
from src.games import TicTacToe
from src.minimax import (
    minimax_decision,
    minimax_decision_with_stats,
    depth_limited_minimax_decision,
    depth_limited_minimax_value,
    depth_limited_minimax_with_stats,
)
from src.evaluation import (
    evaluate_tic_tac_toe,
)


<h1> Tic-Tac-Toe Game </h1> <h3> Step 1: Game Environment Only </h3>

In [2]:
game = TicTacToe()

game

In [3]:
state = game.initial_state()

state

((' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' '), 'X')

In [4]:
game.actions(state)

[0, 1, 2, 3, 4, 5, 6, 7, 8]

In [5]:
next_state = game.result(
    state,
    4
)

next_state

((' ', ' ', ' ', ' ', 'X', ' ', ' ', ' ', ' '), 'O')

<h3> Step 2: MinMax Algorithm </h3>

In [6]:
board1 = (
    "X", "X", " ",
    "O", "O", " ",
    " ", " ", " ",
)

board2 = (
    "X", " ", " ",
    "O", "O", " ",
    "X", " ", " ",
)

board3 = (
    "O", "O", " ",
    "X", "X", " ",
    " ", " ", " ",
)

In [7]:
game = TicTacToe()

state = game.initial_state()

In [8]:
action = minimax_decision(
    game,
    state,
)

action

0

Board Example

In [9]:
game = TicTacToe()

state = (
    board1,
    "X",
)

state

(('X', 'X', ' ', 'O', 'O', ' ', ' ', ' ', ' '), 'X')

In [10]:
best_action = minimax_decision(
    game,
    state,
)

best_action

2

In [11]:
game = TicTacToe()

state = (
    board2,
    "X",
)

state

(('X', ' ', ' ', 'O', 'O', ' ', 'X', ' ', ' '), 'X')

In [12]:
minimax_decision(
    game,
    state,
)

5

<h3> Step 3: Measuring Minimax Search Complexity </h3>

In [13]:
game = TicTacToe()

state = game.initial_state()

state

((' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' '), 'X')

In [14]:
action, stats = minimax_decision_with_stats(
    game,
    state,
)

action

0

In [15]:
stats.summary()

{'Nodes Visited': 549945, 'Terminal States': 255168, 'Max Depth': 9}

A State may be built from different paths.
We don't have Memoization yet.
Some branches will end before the Board is completely filled.

Here we have two options:

1. Restrict the search.</br>
2. Improve the quality of the estimate.

<h3> Step 4: Depth-Limited Minimax </h3>

Evaluation Function: is an approximation of future value.

Utility is only for the end game and is definite. </br>
Evaluation is for the middle game and is approximate.

The Evaluation Function says, given the current situation, how promising is this state?

In [16]:
game = TicTacToe()

state = (
    board3,
    "X",
)

best_action = minimax_decision(
    game,
    state,
)

best_action

2

In [17]:
evaluate_tic_tac_toe(
    game,
    state
)

1

### Minimax Tie Between Multiple Optimal Actions

In some states, Minimax may return an action that is not the most intuitive move, even though the algorithm is working correctly.

The reason is that multiple actions can lead to the same final utility value. For example, in a Tic-Tac-Toe state:

- Action 5 may create an immediate win.
- Action 2 may also guarantee a win after considering the opponent's optimal responses.

Since the current implementation updates the best action only when:

`value > best_value`

the first action with the maximum value is selected. Therefore, if multiple actions have the same utility, the earliest action in the action list is returned.

This behavior is not an error; it is a tie-breaking issue. In practical game agents, additional strategies such as random selection among equally optimal actions or heuristic-based tie-breaking can be used.

In [18]:
game = TicTacToe()

state = game.initial_state()
print(state)

action = depth_limited_minimax_decision(
    game,
    state,
    depth_limit=3,
    evaluation_function=evaluate_tic_tac_toe,
)

action

((' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' '), 'X')


4

**Important point**

In Tic-Tac-Toe, the center is usually the best starting move, as it creates the most potential winning lines.


Full Minimax may initially see every move as equal in terms of final outcome, as the full game is played all optimal moves can produce a Draw. </br>
But Depth-Limited, because it works with Evaluation, sees the differences in heuristic value and chooses the center.

Depth-Limited is dependent on Evaluation Function Quality and if Evaluation is poor, it may choose a wrong move.

In [19]:
game = TicTacToe()

state = (
    board2,
    "X",
)
print(state)

action = depth_limited_minimax_decision(
    game,
    state,
    depth_limit=3,
    evaluation_function=evaluate_tic_tac_toe,
)

action


(('X', ' ', ' ', 'O', 'O', ' ', 'X', ' ', ' '), 'X')


5

In [20]:
game = TicTacToe()

state = (
    board3,
    "X",
)
print(state)

action = depth_limited_minimax_decision(
    game,
    state,
    depth_limit=3,
    evaluation_function=evaluate_tic_tac_toe,
)

action

(('O', 'O', ' ', 'X', 'X', ' ', ' ', ' ', ' '), 'X')


2

The Evaluation Function implemented so far, in this case, does not properly distinguish between immediate win or longer win between moves.

In fact, this test revealed a weakness in the Evaluation Function.

In [21]:
depth_limit = 1
depth_limit = 2
depth_limit = 3
depth_limit = 5

In [22]:
for depth in [1,2,3,5]:
    action = depth_limited_minimax_decision(
        game,
        state,
        depth_limit=depth,
        evaluation_function=evaluate_tic_tac_toe,
    )
    print(depth, action)
    
print("Values:")
for action in game.actions(state):
    value = depth_limited_minimax_value(
        game,
        game.result(
            state,
            action,
        ),
        depth=1,
        depth_limit=1,
        evaluation_function=evaluate_tic_tac_toe,
    )

    print(action, value)

1 5
2 5
3 2
5 2
Values:
2 21
5 10000
6 11
7 2
8 3


Depth-Limited Minimax depends on the quality of the Evaluation Function, and increasing Depth does not necessarily mean choosing the expected human motion; Rather, it means a better decision based on Utility.

<h3> Step 5: Measuring Minimax vs Depth-Limited Minimax Complexity </h3>

Search function → decides</br>
Value function → calculates the value of State

In [23]:
for depth in [1,2,3,5,9]:
    action, stats = depth_limited_minimax_with_stats(
        game,
        state,
        depth_limit=depth,
        evaluation_function=evaluate_tic_tac_toe,
    )
    print(
        "Depth:",
        depth,
        "Action:",
        action,
        stats.summary(),
    )

Depth: 1 Action: 5 {'Nodes Visited': 5, 'Terminal States': 1, 'Max Depth': 1}
Depth: 2 Action: 5 {'Nodes Visited': 21, 'Terminal States': 4, 'Max Depth': 2}
Depth: 3 Action: 2 {'Nodes Visited': 60, 'Terminal States': 19, 'Max Depth': 3}
Depth: 5 Action: 2 {'Nodes Visited': 144, 'Terminal States': 67, 'Max Depth': 5}
Depth: 9 Action: 2 {'Nodes Visited': 144, 'Terminal States': 67, 'Max Depth': 5}


More depth does not always: give better action.

Rather: gives more information about the decision space. </br>
Then our policy makes the decision.

Comparison with Full Minimax

**Full Minimax:**</br>
Nodes Visited = 549945</br>
Terminal States = 255168</br>
Max Depth = 9

**Now:**</br>
Depth 5:</br>
Nodes Visited = 144</br>
Terminal States = 67</br>

That is:</br>
A very drastic reduction.</br>
But since the game is small, Full Minimax is still possible.

### Depth-Limited Minimax Behavior

Depth-Limited Minimax reduces computational cost by stopping the search at a predefined depth and using an evaluation function instead of the exact utility.

Increasing depth generally increases the number of explored nodes. However, if the game reaches terminal states before the depth limit, increasing the limit further does not increase the search space.

Deeper search may also change the selected action because the agent can discover additional optimal paths. When multiple actions have the same final utility, the selected action depends on the tie-breaking strategy.